opensearch-py client and enable llama-index's MetaFilters.  Use appro…

aerickson-clt opened the pull request:
…x-knn when no filters, and use painless-scripting method when there are (pre-)filters.

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #7004

## Type of Change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] I stared at the code and made sure it makes sense
- [ ] I manually tested against my own aws opensearch instance and some test data, similar to the opensearch notebook in llama index.

# Suggested Checklist:

- [ ] I have performed a self-review of my own code

logan-markewich commented on llama_index/vector_stores/opensearch.py:
```diff
@@ -1,16 +1,164 @@
+    return client
+
+
+def _get_kwargs_value(kwargs: Any, key: str, default_value: Any) -> Any:
```
This can be implemented with just `kwargs.get(key, default_value)`

No need for this function

logan-markewich reviewed (commented):
I think this looks good? Is there an easy guide you can recommend for testing this locally though? Maybe updating a notebook for this would be nice

aerickson-clt commented:
I have been trying this on a modified OpenSearchDemo notebook and am encountering an exception.  Probably shouldn't merge yet.

aerickson-clt commented:
I demoed the filters in the OpenSearchDemo notebook and addressed the linting  issues.  I have tested this with my AWS OpenSearch service instance, but not with the docker instance as written in the notebook.

I also modified the filters so that you can do full-featured filter-context queries as documented at https://opensearch.org/docs/latest/query-dsl/query-filter-context/

aerickson-clt commented:
@logan-markewich Hopefully this is ready enough now

Disiok commented:
> @logan-markewich Hopefully this is ready enough now

looks like still some small lint issues 😅  I can try to fix and land this tomorrow morning.
